lifecycle.go

```diff
@@ -59,47 +63,72 @@ type Appender struct {
 // decorators provides a list of optional constructor functions that will return decorators
 // that wrap the base appender. This can be used to provide deduplication. Decorators will be
 // called in-order, and the last in the chain will be the base appender.
-// TODO(mhutchinson): switch the decorators over to a WithOpt for future flexibility.
-func NewAppender(d Driver, opts ...func(*appendOptions)) (*Appender, LogReader, error) {
-	resolved := resolveAppendOpts(opts...)
-	type appender interface {
-		Add(ctx context.Context, entry *Entry) IndexFuture
-		// Shutdown(ctx context.Context)
+func NewAppender(ctx context.Context, d Driver, opts ...func(*AppendOptions)) (*Appender, LogReader, error) {
+	resolved := resolveAppendOptions(opts...)
+	type appendLifecycle interface {
+		Appender(context.Context, *AppendOptions) (*Appender, error)
 	}
-	a, ok := d.(appender)
+	lc, ok := d.(appendLifecycle)
 	if !ok {
-		return nil, nil, fmt.Errorf("driver %T does not implement Appender", d)
+		return nil, nil, fmt.Errorf("driver %T does not implement Appender lifecycle", d)
+	}
+	a, err := lc.Appender(ctx, resolved)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to init appender lifecycle: %v", err)
 	}
-	add := a.Add
-	for i := len(resolved.addDecorators) - 1; i >= 0; i-- {
-		add = resolved.addDecorators[i](add)
+	for i := len(resolved.AddDecorators) - 1; i >= 0; i-- {
+		a.Add = resolved.AddDecorators[i](a.Add)
 	}
 	reader, ok := d.(LogReader)
 	if !ok {
 		return nil, nil, fmt.Errorf("driver %T does not implement LogReader", d)
 	}
-	return &Appender{
-		Add:      add,
-		// Shutdown: a.Shutdown,
-	}, reader, nil
+	return a, reader, nil
 }
 
-func WithAppendDeduplication(decorators ...func(AddFn) AddFn) func(*appendOptions) {
-	return func(o *appendOptions) {
-		o.addDecorators = decorators
+func WithAppendDeduplication(decorators ...func(AddFn) AddFn) func(*AppendOptions) {
+	return func(o *AppendOptions) {
+		o.AddDecorators = decorators
 	}
 }
 
-type appendOptions struct {
-	addDecorators []func(AddFn) AddFn
+// NewCPFunc is the signature of a function which knows how to format and sign checkpoints.
+type NewCPFunc func(size uint64, hash []byte) ([]byte, error)
+
+// ParseCPFunc is the signature of a function which knows how to verify and parse checkpoints.
+type ParseCPFunc func(raw []byte) (*f_log.Checkpoint, error)
+
+// EntriesPathFunc is the signature of a function which knows how to format entry bundle paths.
+type EntriesPathFunc func(n uint64, p uint8) string
+
+// AppendOptions holds optional settings for all storage implementations.
+type AppendOptions struct {
```

Comment:
I'm not immediately convinced that all of these need to be publicly exposed from the API?

Reply:
I've removed the func types for now, we can figure out whether/how to reintroduce in a less polluting way later on.

`AppendOptions` needs to be exported so that storage impls can accept it, but again I think we can return to this and see if there's a better way/place - the primary goal of _this_ PR is just to get us over the mark of broadly adopting the lifecycle API changes as it's blocking other work.